phoenix-core/src/main/java/org/apache/phoenix/coprocessor/GroupedAggregateRegionObserver.java

```diff
@@ -373,198 +389,254 @@ GroupByCache newCache(RegionCoprocessorEnvironment env, ImmutableBytesPtr tenant
+                                new ArrayList<Cell>();
                         // Results are potentially returned even when the return
                         // value of s.next is false
                         // since this is an indication of whether or not there are
                         // more values after the
                         // ones returned
-                        hasMore = scanner.nextRaw(results);
+                        hasMore = delegate.nextRaw(results);
                         if (!results.isEmpty()) {
                             result.setKeyValues(results);
                             ImmutableBytesPtr key =
-                                TupleUtil.getConcatenatedValue(result, expressions);
+                                    TupleUtil.getConcatenatedValue(result, expressions);
                             Aggregator[] rowAggregators = groupByCache.cache(key);
                             // Aggregate values here
                             aggregators.aggregate(rowAggregators, result);
                         }
-                    } while (hasMore && groupByCache.size() < limit);
+                        now = EnvironmentEdgeManager.currentTimeMillis();
+                    } while (hasMore && groupByCache.size() < limit && (now - startTime) < pageSizeInMs);
+                    if (hasMore && groupByCache.size() < limit && (now - startTime) >= pageSizeInMs) {
+                        // Return a dummy result as we have processed a page worth of rows
+                        // but we are not ready to aggregate
+                        getDummyResult(resultsToReturn);
+                        return true;
+                    }
+                    regionScanner = groupByCache.getScanner(delegate);
+                    // Do not sort here, but sort back on the client instead
+                    // The reason is that if the scan ever extends beyond a region
+                    // (which can happen if we're basing our parallelization split
+                    // points on old metadata), we'll get incorrect query results.
+                    return regionScanner.next(resultsToReturn);
                 }
-            }  finally {
+            } finally {
                 if (acquiredLock) region.closeRegionOperation();
             }
+        }
 
-            RegionScanner regionScanner = groupByCache.getScanner(scanner);
-
-            // Do not sort here, but sort back on the client instead
-            // The reason is that if the scan ever extends beyond a region
-            // (which can happen if we're basing our parallelization split
-            // points on old metadata), we'll get incorrect query results.
-            success = true;
-            return regionScanner;
-        } finally {
-            if (!success) {
+        @Override
+        public void close() throws IOException {
+            if (regionScanner != null) {
+                regionScanner.close();
+            } else {
```

Comment:
Good catch! It is needed. I will add it back.

Reply:
Actually, it is not needed any more as I introduced a new variable called regionScanner which is the scanner returned by the cache. Previously, success is set true only when the cache returned the scanner successfully. In this patch, this means regionScanner != null. When regionScanner is not null we call close() on  regionScanner which closes the cache. Otherwise, we close the cache, i.e., Closeables.closeQuietly(groupByCache).